src/main/java/org/jabref/logic/importer/fetcher/GrobidCitationFetcher.java

```diff
@@ -0,0 +1,91 @@
+package org.jabref.logic.importer.fetcher;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.TreeSet;
+
+import org.jabref.logic.importer.FetcherException;
+import org.jabref.logic.importer.ImportFormatPreferences;
+import org.jabref.logic.importer.ParseException;
+import org.jabref.logic.importer.SearchBasedFetcher;
+import org.jabref.logic.importer.fileformat.BibtexParser;
+import org.jabref.logic.importer.util.GrobidService;
+import org.jabref.logic.importer.util.GrobidServiceException;
+import org.jabref.model.entry.BibEntry;
+import org.jabref.model.util.FileUpdateMonitor;
+import org.jabref.preferences.JabRefPreferences;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class GrobidCitationFetcher implements SearchBasedFetcher {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(GrobidCitationFetcher.class);
+    private ImportFormatPreferences importFormatPreferences;
+    private FileUpdateMonitor fileUpdateMonitor;
+    private GrobidService grobidService;
+    private List<String> failedEntries = new ArrayList<>();
+
+    public GrobidCitationFetcher(ImportFormatPreferences importFormatPreferences, FileUpdateMonitor fileUpdateMonitor, JabRefPreferences jabRefPreferences) {
+        this.importFormatPreferences = importFormatPreferences;
+        this.fileUpdateMonitor = fileUpdateMonitor;
+        grobidService = new GrobidService(jabRefPreferences);
+    }
+
+    /**
+     * Passes request to grobid server, using consolidateCitations option to improve result.
+     * Takes a while, since the server has to look up the entry.
+     */
+    private String parseUsingGrobid(String plainText) throws FetcherException {
+        try {
+            return grobidService.processCitation(plainText, 1);
+        } catch (GrobidServiceException e) {
+            throw new FetcherException("The Pipeline failed to get the results from the GROBID client", e);
+        }
+    }
+
+    private Optional<BibEntry> parseBibToBibEntry(String bibtexString) throws FetcherException {
+        try {
+            return BibtexParser.singleFromString(bibtexString,
+                    importFormatPreferences, fileUpdateMonitor);
+        } catch (ParseException e) {
+            throw new FetcherException("Jabref failed to extract a BibEntry form bibtexString.", e);
+        }
+    }
+
+    @Override
+    public List<BibEntry> performSearch(String query) throws FetcherException {
+        TreeSet<String> plainReferences = new TreeSet<>();
+        String[] plainReferencesArray = query.split(";;");
```

Comment:
```
Arrays.stream( query.split( ";;" ) )
          .map(trim)
          .filter(String::notempty)
          .collect(toTreeset);
```
is a bit more readable.

Reply:
Resolved here: d6aec800e84236fc9612dcd37a07bfa8f7602ee4